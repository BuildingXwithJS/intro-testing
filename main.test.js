/* eslint-env jest */
// npm packages
const nock = require('nock');
const sinon = require('sinon');
const inquirer = require('inquirer');

// our packages
const {getData, handler} = require('./main');

const gamescomCrate = {
  id: '578080',
  name: 'GAMESCOM INVITATIONAL CRATE',
};

// mock opn function
jest.mock('opn');
const opn = require('opn');

describe('Main function', () => {
  let steamcommunity;

  beforeAll(() => {
    const url =
      `/market/priceoverview/?appid=${gamescomCrate.id}&` +
      `country=DE&currency=3&market_hash_name=${encodeURIComponent(gamescomCrate.name)}`;
    steamcommunity = nock('http://steamcommunity.com')
      .get(url)
      .times(3)
      .reply(200, {
        lowest_price: '6,10€',
        median_price: '6,10€',
        success: true,
        volume: '8,472',
      });
  });

  afterAll(() => {
    steamcommunity.cleanAll();
  });

  test('Gets default data', async done => {
    const result = await getData(gamescomCrate);
    expect(result).toMatchSnapshot();
    done();
  });

  test('Renders correct data to a user', async done => {
    // stub the reply from user
    const inquirerStub = sinon.stub(inquirer, 'prompt').callsFake(() => ({openBrowser: false}));
    // tap into console.log function
    const consoleSpy = sinon.spy(console, 'log');
    // execute main login
    await handler(gamescomCrate);
    // remove console spy
    consoleSpy.restore();
    // validate output
    expect(consoleSpy.args).toMatchSnapshot();
    // restore inquirer
    inquirerStub.restore();
    done();
  });

  test('Opens browser on user request', async done => {
    // stub the reply from user
    const inquirerStub = sinon.stub(inquirer, 'prompt').callsFake(() => ({openBrowser: true}));
    // tap into console.log function
    const consoleSpy = sinon.spy(console, 'log');
    // execute main login
    await handler(gamescomCrate);
    // remove console spy
    consoleSpy.restore();
    // validate output
    expect(consoleSpy.args).toMatchSnapshot();
    // make sure opn was called
    expect(opn).toBeCalled();
    // restore inquirer
    inquirerStub.restore();
    done();
  });
});
