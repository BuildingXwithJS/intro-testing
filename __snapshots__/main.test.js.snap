// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main function Gets default data 1`] = `
Object {
  "lowest_price": "6,10â‚¬",
  "median_price": "6,10â‚¬",
  "success": true,
  "volume": "8,472",
}
`;

exports[`Main function Opens browser on user request 1`] = `
Array [
  Array [
    "Right now [1mGamescom Invitational Crate[22m costs:
    [90m- Lowest price:[39m [32m6,10â‚¬[39m
    [90m- Median price:[39m [34m6,10â‚¬[39m
    [90m- Volume:[39m [33m8,472[39m",
  ],
]
`;

exports[`Main function Renders correct data to a user 1`] = `
Array [
  Array [
    "Right now [1mGamescom Invitational Crate[22m costs:
    [90m- Lowest price:[39m [32m6,10â‚¬[39m
    [90m- Median price:[39m [34m6,10â‚¬[39m
    [90m- Volume:[39m [33m8,472[39m",
  ],
]
`;
